ldr_ro: convert to ServiceFramework

wwylele opened the pull request:
Also fixes all `process` usage, and creates storage for each client.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/citra-emu/citra/3192)
<!-- Reviewable:end -->

Subv commented:
Reviewed 3 of 3 files at r1, 2 of 2 files at r2, 4 of 4 files at r3.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/citra-emu/citra/3192)*
<!-- Sent from Reviewable.io -->

lioncash commented on src/core/hle/service/ldr_ro/memory_synchronizer.h:
```diff
@@ -5,6 +5,7 @@
 #pragma once
 
 #include <vector>
+#include "core/hle/kernel/process.h"
```
Forward declaration?

lioncash commented on src/core/hle/service/ldr_ro/ldr_ro.h:
```diff
@@ -4,19 +4,156 @@
+     *      0 : Return header
+     *      1 : Result of function, 0 on success, otherwise error code
+     */
+    void Initialize(Kernel::HLERequestContext& ctx);
```
As far as I can tell, the functions can also be made private, unless I'm mistaken, as everything goes through the internal function table.

lioncash commented on src/core/hle/service/ldr_ro/ldr_ro.h:
```diff
@@ -4,19 +4,156 @@
 #pragma once
 
+#include <unordered_map>
+#include "core/hle/kernel/kernel.h"
```
Unless I'm mistaken, this can be a forward declaration, despite the LoadCRO template, as it's passing a reference to another function taking a reference (so the reference isn't dereferenced).

wwylele replied:
actually I don't even need the declaration because it must have been declared by the service framework... will just remove this

wwylele commented:
addressed

wwylele commented:
Removed canary merge tag and let #3227 go into test first.

B3n30 commented on src/core/hle/service/ldr_ro/ldr_ro.cpp:
```diff
@@ -582,44 +490,36 @@ static void LinkCRO(Interface* self) {
+    if (find_result == slots.end()) {
         LOG_ERROR(Service_LDR, "Not initialized");
         rb.Push(ERROR_NOT_INITIALIZED);
+        rb.Push<u32>(0);
```
Adding an additional normal parameter, so IPC::RequestBuilder rb = rp.MakeBuilder(1, 0); must be different for that case
Why is this parameter even necessary?

wwylele replied:
oops, copy-paste error

B3n30 commented on src/core/hle/service/ldr_ro/ldr_ro.cpp:
```diff
@@ -635,83 +535,72 @@ static void UnlinkCRO(Interface* self) {
+    if (find_result == slots.end()) {
         LOG_ERROR(Service_LDR, "Not initialized");
         rb.Push(ERROR_NOT_INITIALIZED);
+        rb.Push<u32>(0);
```
Same as above

wwylele commented:
(Hold on. need to fix some stuff)

wwylele commented:
Rebased and addressed comments. Also added a TODO about passing process handle to CRO helper. I didn't finish this because CRO helper uses lots of Read/Write 8/16/32 which haven't got their overload with process parameter. I don't want to put too many change in one PR
